File: Solution0002.cs
/**
 Uploaded: 10/03/2021 16:14
 Runtime: 100 ms, 74.10%
 Memory Usage: 28.3 MB, 79.99%
 */

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        int[] resultValue = new int[100];

        int count = 0;

        ListNode result = null;

        while (true)
        {
            if (l1 != null)
            {
                resultValue[count] += l1.val;
                l1 = l1.next;
            }
            if (l2 != null) { 
                resultValue[count] += l2.val;
                l2 = l2.next;
            }

            if (resultValue[count] > 9)
            {
                resultValue[count + 1] = resultValue[count] / 10;
                resultValue[count] = resultValue[count] % 10;
            }

            count++;

            if (l1 == null && l2 == null)
            {
                if(resultValue[count] > 0) count++;
                break;
            }
        }

        for(int index = count-1; index >= 0; index--)
        {
            result = new ListNode(resultValue[index], result);
        }

        return result;
    }
}

File: Rust/solution0002.rs
/*
Nov 23, 2024 16:12
Runtime 0ms, Beats 100.00%
Memory 2.28MB, Beats 16.67%
*/

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn add_two_numbers(
        l1: Option<Box<ListNode>>,
        l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut dummy_head = Box::new(ListNode::new(0));
        let mut curr = &mut dummy_head;
        let mut carry = 0;

        let (mut l1, mut l2) = (l1, l2);

        while l1.is_some() || l2.is_some() || carry != 0 {
            let x = l1.as_ref().map_or(0, |node| node.val);
            let y = l2.as_ref().map_or(0, |node| node.val);
            let sum = carry + x + y;
            carry = sum / 10;

            curr.next = Some(Box::new(ListNode::new(sum % 10)));
            curr = curr.next.as_mut().unwrap();

            l1 = l1.and_then(|node| node.next);
            l2 = l2.and_then(|node| node.next);
        }

        dummy_head.next
    }
}
